### ISSUE TITLE
Add PropNameID::fromSymbol to JSI

### ISSUE THREAD

**Issue by TatianaKapos** [microsoft/react-native-windows#9703]
### Problem Description

The lastest integrations https://github.com/facebook/react-native/compare/7cece3423...189c2c895 added the ability to access properties with symbol keys through JSI https://github.com/facebook/react-native/commit/9010bfe457b77862024214ce6210504ff1786ef5. This broke us and we need to add these functions to the windows-side to get rid of the overrides (may also need to add the function to hermes-windows)

Errors:

`C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\include\memory(3415,0): Error C2259: Microsoft::JSI::anonymous-namespace'::NapiJsiRuntime': cannot instantiate abstract class`

`C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\include\memory(3416,35): Error C2259: 'Microsoft::JSI::SystemChakraRuntime': cannot instantiate abstract class`

	Assertion failed!
	
	Program: ...round\windows\x64\Debug\playground\AppX\hermes.dll
	File: D:\a\1\s\include\hermes/VM/HermesValue.h
	Line: 409
	
	Expression: isSymbol()
	
	For information on how your program can cause an assertion
	failure, see the Visual C++ documentation on asserts
	
	(Press Retry to debug the application - JIT must be enabled)playground.exe has triggered a breakpoint.
	
	Assertion failed!
	
	Program: ...round\windows\x64\Debug\playground\AppX\hermes.dll
	File: D:\a\1\s\include\hermes/VM/SymbolID.h
	Line: 116
	
	Expression: id <= LAST_INVALID_ID && "ID outside valid range"
	
	For information on how your program can cause an assertion
	failure, see the Visual C++ documentation on asserts
	
	(Press Retry to debug the application - JIT must be enabled)playground.exe has triggered a breakpoint.
	
	Assertion failed!
	
	Program: ...round\windows\x64\Debug\playground\AppX\hermes.dll
	File: D:\a\1\s\lib\VM\DictPropertyMap.cpp
	Line: 97
	
	Expression: symbolID.isValid() && "looking for an invalid SymbolID"
	
	For information on how your program can cause an assertion
	failure, see the Visual C++ documentation on asserts

### Steps To Reproduce

1. Remove override files (JSCRuntime.cpp, Jsi.h, decorator.h )
2. Run the playground app

### Expected Results

Playground to run w/o issues

### CLI version

npx react-native --version

### Environment

```markdown
System:
    OS: Windows 10 10.0.22000
    CPU: (8) x64 Intel(R) Core(TM) i7-9700K CPU @ 3.60GHz
    Memory: 49.85 GB / 63.85 GB
  Binaries:
    Node: 14.17.5 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.5 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 8.2.0 - C:\Program Files\nodejs\npm.CMD
    Watchman: Not Found
  SDKs:
    Android SDK: Not Found
    Windows SDK:
      AllowDevelopmentWithoutDevLicense: Enabled
      AllowAllTrustedApps: Enabled
      Versions: 10.0.17763.0, 10.0.18362.0, 10.0.19041.0, 10.0.22000.0
  IDEs:
    Android Studio: Not Found
    Visual Studio: 16.11.32126.315 (Visual Studio Enterprise 2019)
  Languages:
    Java: 1.8.0_302 - C:\Program Files\Eclipse Foundation\jdk-8.0.302.8-hotspot\bin\javac.EXE
  npmPackages:
    @react-native-community/cli: Not Found
    react: Not Found
    react-native: Not Found
    react-native-windows: Not Found
  npmGlobalPackages:
    *react-native*: Not Found
```

### Target Platform Version

10.0.19041

### Target Device(s)

Desktop

### Visual Studio Version

Visual Studio 2019

### Build Configuration

Debug

**Comment by asklar** [comment]
curious how this is handled for other platforms? in that RN commit it shows throwing an exception for JSC

**Comment by chrisglein** [comment]
Likely requires some kind of paired change from Hermes. Does that sound right @aeulitz?

Currently this was handled in the integration by providing a stub function.

**Comment by TatianaKapos** [comment]
Some more context: 

We have a stub functions in **JsiAbiApi.cpp, JsiAbiApi.h, NodeApiJsiRuntime** and **overrides in ReactCommon** to fix this. The overrides in ReactCommon fixes the Hermes errors.

I _think_ to fully remove the overrides we would have to add stub functions (or implementations) to ChakraRuntime, and hermes-windows

**Comment by NickGerleman** [comment]
Please note that locally modifying the JSI API will break adoption of future versions of Hermes and V8. I you haven't already, I'd recommend talking with the JSHost folks who work in this area.

**Comment by NickGerleman** [comment]
Historically, I've talked to @mganandraj and @tudorms for Hermes and V8 when we needed JS engine releases using newer ABI targets, but I'm not sure who's responsible for that these days.

**Comment by NickGerleman** [comment]
There are a couple workstreams that help to address the issue, in different ways. @vmoroz has been looking at creating a stable J JS engine ABI, which could allow looser interoperability between different versions of RN, and JS engine binaries. There have been ideas going in the opposite direction, for Hermes to be distributed as part of RN itself, making the ABI consistent between projects as they are built. @neilDhar or @cortinico might know more about the thoughts on that one.

**Comment by NickGerleman** [comment]
Related: https://github.com/facebook/hermes/issues/706

**Comment by cortinico** [comment]
> There have been ideas going in the opposite direction, for Hermes to be distributed as part of RN itself, making the ABI consistent between projects as they are built. @neildhar or @cortinico might know more about the thoughts on that one.

Following up here, I'd like to point out on this comment from the New Architecture working group:

- https://github.com/reactwg/react-native-new-architecture/discussions/4#discussioncomment-2382960

Specifically, we ended up implementing the second point that @NickGerleman mentioned. We're bundling a Hermes version alongside React Native. This essentially eliminates the ABI stability problem as framework & engine are built at the same time.

Hope this helps clarify the problem and the potential solution.

**Comment by tudorms** [comment]
@cortinico how do we pick which commit of Hermes goes with which commit of RN? Will we have RN tags in the Hermes repo?

Will JSI still be considered a public API at that point (does it have versioning guarantees, etc.)?

**Comment by tudorms** [comment]
It sounds like the ABI issue will still be there if extensions will continue using JSI as a public API. Unless we're saying everyone needs to build everything from source (framework, engine, and all extensions), or extension authors need to produce new binary versions for each rev of the RN version?

**Comment by vmoroz** [comment]
What about using other JS engines? We have valid scenarios for using V8 in some desktop applications. If there will be no well defined API between JS engine and RN, then we cannot support them.
Would Meta engineers be opened to discussing establishing a well defined API?
Who would need the JSI public API if it is not used by RN?

**Comment by neildhar** [comment]
I defer to @cortinico on the question about building extensions from source, since that's independent of Hermes.

RN still intends to support other JS engines, they can continue to be versioned independently of RN, and accessed through JSI. This is an organisational change to avoid ABI incompatibility issues with Hermes, and improve our ability to respond to release issues, as we work towards making it the default. This is conceptually similar to how the JSI layer for JSC currently lives in the RN codebase and is built together with RN.

**Comment by cortinico** [comment]
Hey all,
Happy to clarify a couple of points here:

> I defer to @cortinico on the question about building extensions from source, since that's independent of Hermes.

The change we announced https://github.com/reactwg/react-native-new-architecture/discussions/4 is not going to affect consumer of developers in any form.

Historically, we used to cut a release of Hermes for a specific version of RN.
For instance: https://github.com/facebook/hermes/releases/tag/v0.11.0
was version `v0.11.0 for RN0.68.x`. 

In the new approach, we will be tagging a version of Hermes once we cut a release branch of RN (with this GH Action https://github.com/facebook/hermes/blob/main/.github/workflows/create-tag.yml).

This tag name will then be stored inside the `sdks/.hermesversion` file inside RN. We will then be using this tag to download the Hermes source code and build it for that specific release branch. This is essentially the synchronization point between the two repo. Moreover, this approach allows us to tag a specific version of Hermes, should we need a patch once we're inside the RN release process.

> Unless we're saying everyone needs to build everything from source (framework, engine, and all extensions), or extension authors need to produce new binary versions for each rev of the RN version?

We're not saying that everyone needs to build everything from source. We're saying that we're distributing RN with a version of Hermes that was built from source at release time.

If you're building on top of Hermes, you can rely on those tags. Otherwise, you can grab a prebuilt Hermes version bundled inside RN.

> @vmoroz @tudorms I've asked that you receive permissions to contribute to RN discussion in reactwg GitHub discussions.

Thanks for doing that @NickGerleman.
If someone else wish to have access to the WG, you can do so with the form on the README of that repo.

> I do not think this change has implications on Windows ability to distribute Hermes as a binary,

I agree on this. I'm missing on how this change is impacting you directly.
I can understand that all the tagging and release workflow was missing in my initial post (as the target audience was RN users) so this could have created some confusion. I'll make sure to update that.

I hope this clarifies.

Also, there were several questions raised on JSI API, which I believe @neildhar is the best POC to provide answers on.

Finally, I think this conversation should be happening inside either the WG or another broader channel (and on a specific issue) as it will be harder to crawl and find in the future.

**Comment by NickGerleman** [comment]
> Otherwise, you can grab a prebuilt Hermes version bundled inside RN.

React Native for Windows has been using a Microsoft created binary distribution of Hermes, instead of the one provided by Meta. E.g. the MSFT distribution is being used to test JS engine ABIs that are being standardized by the NodeJS team. Some changes that may look internal indirectly end up impacting them downstream. Both as creators of an in-production Hermes implementation, and consumers of Hermes. 

Downstream of RNW is other long-running forks. E.g. Messenger has been public about maintaining a Meta internal fork which upstream OSS RNW (but they do their own Hermes distribution). So there is a Messenger version of MSFT RNW doing special things we don't expect sometimes🙂. There's a hidden web, often with different compliance, build reliability requirements.

**Comment by tudorms** [comment]
Thank you for the clarification. Might it be worth making Hermes a submodule of react-native?

We're interested in this from 2 perspectives:
* wanting to align in React Native Windows with how Hermes is built / integrated on other platforms (considering we're doing an ugly hack for the Hermes debugger build where we need react native code in the Hermes repo, a tighter alignments makes sense either way);
* making sure JSI doesn't become an internal detail; it's easier / more natural to make spanning changes between Hermes and React Native if JSI is seen as an internal layer between the 2, but not properly versioning JSI breaks integration with external engines (we use [V8](https://github.com/microsoft/v8-jsi) in some scenarios in Office) and breaks the binary distribution story for JSI/TurboModule-based extensions.

**Comment by NickGerleman** [comment]
Re channel, a goal I would have is to prioritize access to implementors of RN and JS engine binaries. I think `reactwg` can be that place, if it solicits access to the broad community building the RN platform. That might be a call for Office CXE folks doing the behind the scenes work for React Native Windows implementation to join these.

**Comment by NickGerleman** [comment]
@cortinico I would recommend adding partners of RN preemptively to `reactwg`, otherwise discussion is likely to continue in the distinct places where individuals have access.

**Comment by NickGerleman** [comment]
Disclaimer that I no longer work at Microsoft, and no longer work on RNW day to day, but when I worked on it, I worked closely with other folks at MSFT folks on Hermes production-readiness. I have no obligation or incentives around this, but I could imagine some really cool outcomes of the change for the RN desktop platform.

**Comment by NickGerleman** [comment]
@Khalef1 could you follow up with the instructions provided on adding appropriate folks to the `reactwg` group

**Comment by NickGerleman** [comment]
@lyahdav @rozele or @vmoroz might be able to provide additional context into JS engine injection goals. The RNW solution for users of Hermes DLL was IIRC going to be NodeApi-based, with JSI bridge for TurboModules and other JSI extensions. But I know we also wanted to provide a way to link Hermes into the same DLL as the application, for reasons like binary deduplication, or the ability to do better optimization.

**Comment by cortinico** [comment]
> Might it be worth making Hermes a submodule of react-native?

In a sense, the current setup is similar to a git submodule. The reason why we haven't decided for this approach are several:

- React Native is backed by an internal mercurial repo, therefore working with git submodules is suboptimal there.
- We wanted to be more explicit with tags like https://github.com/cortinico/hermes/releases/tag/hermes-2022-02-21-RNv0.68.0-rc1-0172d30ac14e8c936c4fd2c435b799e0009aeb00 which are a bit more informative on which version of Hermes source got picked an used for a RN build.
- We wanted to have a way to make sure that the main build of RN runs against a main build of Hermes.

> making sure JSI doesn't become an internal detail; it's easier / more natural to make spanning changes between Hermes and React Native if JSI is seen as an internal layer between the 2, but not properly versioning JSI breaks integration with external engines (we use [V8](https://github.com/microsoft/v8-jsi) in some scenarios in Office) and breaks the binary distribution story for JSI/TurboModule-based extensions.

We're not changing the JSI versioning so in a sense, nothing is changing from the status quo. @neildhar can tell you more on this specifically.

**Comment by NickGerleman** [comment]
I am curious of the thought for NodeAPI JS engine ABI versioning. Basically, if JSI makes a change that is API-breaking, how would we handle it in the ABI-standard layer? Do we expect future JSI changes to follow existing pattern of new methods, structures, etc?

I think it is okay, if we continue to act gracefully when a `jsi::runtime` implementation does not implement new part of JSI spec immediately. The NodeAPI implementation shims ABI incompatibilities, but presumably we'd be working with a smaller set of available APIs when older binaries are used.

**Comment by jonthysell** [comment]
@TatianaKapos Is this already complete? I see the PR https://github.com/microsoft/hermes-windows/pull/94 in hermes-windows, is there a follow-up PR that needs happen here to bump hermes and fix this in RNW? If so, could this make 0.69, or should it be moved to 0.70?
